frontend/lib/uast/BuilderResult.cpp

```diff
@@ -190,90 +190,154 @@ ID BuilderResult::idToParentId(ID id) const {
 
 void BuilderResult::serialize(std::ostream& os) const {
   Serializer ser(os);
-  ser.write(magic);
-  ser.write(version);
+  serialize(ser);
+}
+
+// BuilderResult serialization format (whitespace not significant):
+// <start sentinel string, std::string>
+// <file path, std::string>
+// N: <number of top-level uAST expressions, uint32_t>
+//   0..N-1: <top level expression and its children, AstNode>
+// <id-to-parent-id, llvm::DenseMap<ID, ID>>
+// M: <number of id-to-location entries, uint64_t>
+// <file path for locations, std::string>
+//   0..M-1:
+//     <id as key, ID>
+//     <first line, int>
+//     <first column, int>
+//     <last line, int>
+//     <last column, int>
+// <comment-id-to-location, std::vector<Location>>
+// <end sentinel string, std::string>
+void BuilderResult::serialize(Serializer& ser) const {
+  ser.write(DYNO_BUILDER_RESULT_START_STR);
+  ser.write(filePath_);
   const uint32_t numEntries = numTopLevelExpressions();
   ser.write(numEntries);
 
   for (auto ast : topLevelExpressions()) {
     ast->serialize(ser);
   }
+
+  ser.write(idToParentId_);
+
+  {
+    // TODO: For config-vars set on the command line, the 'file path' is set to
+    // something different. This leads to issues when comparing against a
+    // deserialized BuilderResult.
+    ser.write((uint64_t)idToLocation_.size());
+    ser.write(filePath_);
+    for (const auto& pair : idToLocation_) {
+      ser.write(pair.first);
+      ser.write(pair.second.firstLine());
+      ser.write(pair.second.firstColumn());
+      ser.write(pair.second.lastLine());
+      ser.write(pair.second.lastColumn());
+    }
+  }
+
+  ser.write(commentIdToLocation_);
+  ser.write(DYNO_BUILDER_RESULT_END_STR);
 }
 
-// TODO: handle Locations
-AstList BuilderResult::deserialize(Context* context, std::string& sfname) {
-  std::ifstream myFile;
-  myFile.open(sfname, std::ios::in | std::ios::binary);
+static void assignIDsFromTree(llvm::DenseMap<ID, const AstNode*>& idToAst,
+                              const AstNode* node) {
+  if (node->isComment()) return;
 
-  return deserialize(context, myFile);
+  idToAst[node->id()] = node;
+  for (auto child : node->children()) {
+    assignIDsFromTree(idToAst, child);
+  }
 }
 
-AstList BuilderResult::deserialize(Context* context, std::istream& is) {
-  AstList ret;
-  Deserializer des(context, is);
+BuilderResult BuilderResult::deserialize(Deserializer& des) {
+  BuilderResult ret;
+  AstList alist;
 
-  auto m = des.read<uint64_t>();
-  auto v = des.read<uint32_t>();
-  (void)m; // silence unused variable warnings
-  (void)v; // silence unused variable warnings
-  assert(m == magic);
-  assert(v == version);
+  CHPL_ASSERT(DYNO_BUILDER_RESULT_START_STR == des.read<std::string>());
+
+  auto path = des.read<UniqueString>(); // path
 
   const auto numEntries = des.read<uint32_t>();
 
+  // TODO: for improved performance, try recomputing the IDs rather than
```

Comment:
It seems like we could just compute the IDs using the same `doAssignIds` walk that is done initially after parsing, no? Though this TODO is probably saying as much 😄. At least, I was expecting to see that call, but maybe it's more complicated than that.

Reply:
Right now that ID computation logic is only for a ``Builder`` type, so we'd need to figure out how to make that available for ``BuilderResult``, or consider serializing/deserializing ``Builder``s instead.

Reply:
Ahhhhhh. Hmmmm, you're right that that makes it a bit more difficult.